### ISSUE TITLE
Error 139 When Installing Python 3.6 on Fedora 34

### ISSUE THREAD

**Issue by theSdev** [pyenv/pyenv#1889]
Too many issues will kill our team's development velocity, drastically.
Make sure you have checked all steps below.

### Prerequisite
* [x] Make sure your problem is not listed in [the common build problems](https://github.com/pyenv/pyenv/wiki/Common-build-problems).
* [x] Make sure no duplicated issue has already been reported in [the pyenv issues](https://github.com/pyenv/pyenv/issues). You should look for closed issues, too.
* [x] Make sure you are not asking us to help solving your specific issue.
  * GitHub issues is opened mainly for development purposes. If you want to ask someone to help solving your problem, go to some community site like [Gitter](https://gitter.im/yyuu/pyenv), [StackOverflow](https://stackoverflow.com/questions/tagged/pyenv), etc.
* [x] Make sure your problem is not derived from packaging (e.g. [Homebrew](https://brew.sh)).
  * Please refer to the package documentation for the installation issues, etc.
* [x] Make sure your problem is not derived from plugins.
  * This repository is maintaining `pyenv` and the default `python-build` plugin only. Please refrain from reporting issues of other plugins here.

### Description
- [x] Platform information (e.g. Ubuntu Linux 16.04): Fedora 34
- [x] OS architecture (e.g. amd64): amd64
- [x] pyenv version: 1.2.27
- [x] Python version: I have 3.9.2, wanna install 3.6.12
- [x] C Compiler information (e.g. gcc 7.3): cc (GCC) 11.0.1 20210324 (Red Hat 11.0.1-0)
- [x] Please attach verbose build log as gist
  * You can turn on verbose debug logging using by setting `PYENV_DEBUG=1`, e.g. `env PYENV_DEBUG=1 pyenv install -v 3.6.4`

This is a fresh Fedora 34 installation. I've followed the instructions carefully and installed the prerequisite packages. When I run `pyenv install 3.6.12` I get this error:
```
Last 10 log lines:
(cd /home/thesdev/.pyenv/versions/3.6.12/share/man/man1; ln -s python3.6.1 python3.1)
if test "xupgrade" != "xno"  ; then \
	case upgrade in \
		upgrade) ensurepip="--upgrade" ;; \
		install|*) ensurepip="" ;; \
	esac; \
	 ./python -E -m ensurepip \
		$ensurepip --root=/ ; \
fi
make: *** [Makefile:1102: install] Error 139
```

Link to the verbose log gist: https://gist.github.com/theSdev/9e0be1495f5ba18306c5012a3157ec1d

**Comment by native-api** [comment]
The gist is incomplete.

**Comment by native-api** [comment]
The output you gave is incomplete. It should be a complete output of the `pyenv` command in debug mode as instructed in the issue template.

**Comment by theSdev** [comment]
> The output you gave is incomplete. It should be a complete output of the `pyenv` command in debug mode as instructed in the issue template.

It's not incomplete, the default view is truncated. You need to view it in raw mode to see the full file as is mentioned on top of the page.

**Comment by native-api** [comment]
I saw it in raw mode and it's still incomplete. The trace from `fetch_tarball` to Make invocation is missing.

In any case, prepend the command with `MAKE_OPTS="--trace"` and show the log. This way, we should see what exact command is failing.

**Comment by anton-petrov** [comment]
Based on what I was able to find out, this is pyenv specific behavior in Fedora 34.

Fedora 34:
pyenv install 3.6.* - failed
pyenv install 3.7.* - ok
Ubuntu 20.04:
pyenv install 3.6.* - ok

@native-api 

> prepend the Pyenv invocation with MAKE_OPTS="--trace"

does nothing.

Instead of this `--debug` or `PYENV_DEBUG=1` (which is the same) should be used.

**Comment by anton-petrov** [comment]
So... I found what's the problem - **segmentaion fault**.
Steps to reproduce (build cpython manually):
```
cd ~
mkdir temp
wget https://www.python.org/ftp/python/3.6.13/Python-3.6.13.tgz
tar xzf Python-3.6.13.tgz
cd Python-3.6.13
./configure --enable-optimizations --with-ensurepip=install
make -j8
cd ~/temp/Python-3.6.13

./python 
    Python 3.6.13 (default, May  6 2021, 10:40:25)
    [GCC 11.1.1 20210428 (Red Hat 11.1.1-1)] on linux

./python -E -m ensurepip
Segmentation fault.
```

@theSdev Try an older gcc compiler v10, perhaps this will help you.

**Comment by theSdev** [comment]
Wait, so is this a workaround or solution? No action on `pyenv`'s front is required?

**Comment by native-api** [comment]
> > prepend the Pyenv invocation with MAKE_OPTS="--trace"
> 
> does nothing.

Huh? Shouldn't it be [picked up by Python-build and added to Make invocation](https://github.com/pyenv/pyenv/blob/328fd42c3a2fbf14ae46dae2021a087fe27ba7e2/plugins/python-build/bin/python-build#L752-L756)?

**Comment by stsewd** [comment]
I was able to install 3.6 using clang.

```bash
sudo dnf install clang
CC=clang pyenv install 3.6.13
```

**Comment by dmayilyan** [comment]
Had the same issue with installation of python 3.7.3 and this patch fixed the problem.

**Comment by barunslick** [comment]
Tried every solution I could find but didn't  work. Finally worked after applying that patch. Thanks

**Comment by wizpig64** [comment]
The patched solved building 3.6.11 to .14 for me in a fresh Manjaro vm.

**Comment by xuongtrantrieu** [comment]
F***ing awesome mate!!!

**Comment by zobayer1** [comment]
Any way to do this without using pyenv?

**Comment by dmitriyshashkin** [comment]
@ni-balexand you've saved my day! much appreciated

**Comment by digglife** [comment]
It seems that this patch doesn't work with 3.6.15.

**Comment by dmitriyshashkin** [comment]
@digglife I was able to install 3.6.15 using this patch

**Comment by digglife** [comment]
> @digglife I was able to install 3.6.15 using this patch

Thank you for confirmation. I accidentally added the `diff` command in the beginning of the patch file, so it didn't work. Now it's OK.

```
diff -r 388d58a5de9e Include/objimpl.h
```

**Comment by iongion** [comment]
@ni-balexand That is just great, saved the day, thank you!
Couldn't `pyenv` istelf take care of this for easier user experience ? (maybe / maybe not - but a patch database could really be useful)
Patches like this are also needed for example on Mac OS BigSur+

**Comment by zoj613** [comment]
> I was able to install 3.6 using clang.
> 
> ```shell
> sudo dnf install clang
> CC=clang pyenv install 3.6.13
> ```

This is the only solution that worked for me. Thanks!

**Comment by pavkam** [comment]
> > I was able to install 3.6 using clang.
> > ```shell
> > sudo dnf install clang
> > CC=clang pyenv install 3.6.13
> > ```
> 
> This is the only solution that worked for me. Thanks!

Worked for me as well on latest Manjaro.

**Comment by brandon82890** [comment]
Thank you!  This patch worked fine to install python 3.6.6 with pyenv on Arch Linux.

**Comment by jteppinette** [comment]
I resolved this on Pop!_OS / Ubuntu 21.10 by installing with clang:

```sh
$ sudo apt install clang -y
$ CC=clang pyenv install 3.6.15
```

**Comment by skryvets** [comment]
Pop!_OS / Ubuntu 21.10 user here. Pre-installing these packages helped me to fix the issue:

```bash
sudo apt-get install -y make build-essential libssl-dev zlib1g-dev \
libbz2-dev libreadline-dev libsqlite3-dev wget curl llvm libncurses5-dev \
libncursesw5-dev xz-utils tk-dev libffi-dev liblzma-dev
```

Source - [realpython.com](https://realpython.com/intro-to-pyenv/)

**Comment by titoelfoly** [comment]
in Ubuntu 22.
i had to install gcc-10
then
CC=gcc-10 pyenv install 3.6.13

**Comment by amancontify** [comment]
$ pyenv install --patch 3.6.5 < alignment.patch
After running this, I get the following error, am I doing something wrong?

Installing Python-3.6.5...
can't find file to patch at input line 3
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|--- a/Include/objimpl.h	Tue Sep 06 15:58:40 2016 -0700
|+++ b/Include/objimpl.h	Tue Sep 06 17:11:43 2016 -0700
--------------------------



No file to patch.  Skipping patch.
1 out of 1 hunk ignored
can't find file to patch at input line 15
Perhaps you used the wrong -p or --strip option?
The text leading up to this was:
--------------------------
|diff -r 388d58a5de9e Objects/obmalloc.c
|--- a/Objects/obmalloc.c	Tue Sep 06 15:58:40 2016 -0700
|+++ b/Objects/obmalloc.c	Tue Sep 06 17:11:43 2016 -0700
--------------------------
No file to patch.  Skipping patch.
1 out of 1 hunk ignored

BUILD FAILED (Fedora Linux 35 using python-build 20180424)

Inspect or clean up the working tree at /tmp/python-build.20220623162516.202786
Results logged to /tmp/python-build.20220623162516.202786.log

**Comment by diegogonzalesdesouza** [comment]
> You need to patch some Python 3.6 object allocation code to use 16 byte alignment. It appears that this was fixed in Python 3.7. https://bugs.python.org/issue27987 https://bugs.python.org/file44413/alignment.patch Had to tweak the patch to not use `a` and `b` parent directory names before feeding it to `pyenv`. Tested on Python 3.6.8 because that's what I needed.
> 
> ```
> $ cat alignment.patch
> ```
> 
> ```
> --- Include/objimpl.h
> +++ Include/objimpl.h
> @@ -250,7 +250,7 @@
>          union _gc_head *gc_prev;
>          Py_ssize_t gc_refs;
>      } gc;
> -    double dummy;  /* force worst-case alignment */
> +    long double dummy;  /* force worst-case alignment */
>  } PyGC_Head;
> 
>  extern PyGC_Head *_PyGC_generation0;
> --- Objects/obmalloc.c
> +++ Objects/obmalloc.c
> @@ -643,8 +643,8 @@
>   *
>   * You shouldn't change this unless you know what you are doing.
>   */
> -#define ALIGNMENT               8               /* must be 2^N */
> -#define ALIGNMENT_SHIFT         3
> +#define ALIGNMENT               16               /* must be 2^N */
> +#define ALIGNMENT_SHIFT         4
> 
>  /* Return the number of bytes in size class I, as a uint. */
>  #define INDEX2SIZE(I) (((uint)(I) + 1) << ALIGNMENT_SHIFT)
> ```
> 
> ```
> $ pyenv install --patch 3.6.8 < alignment.patch
> ```
> 
> ```
> Downloading Python-3.6.8.tar.xz...
> -> https://www.python.org/ftp/python/3.6.8/Python-3.6.8.tar.xz
> Installing Python-3.6.8...
> patching file Include/objimpl.h
> Hunk #1 succeeded at 255 (offset 5 lines).
> patching file Objects/obmalloc.c
> Hunk #1 succeeded at 650 (offset 7 lines).
> Installed Python-3.6.8 to /home/asdf/.pyenv/versions/3.6.8
> ```

Just adding to this answer (which solved the problem for me, thanks):
- I had to create a alignment.patch file with `cat > alignment.patch`
- Then, install the Python version I wanted, in my case it was 3.7.0, with `pyenv install --patch 3.7.0 < alignment.patch`
